### PR TITLE
feat: Add `dask_awkward.to_dataframe`

### LINKED ISSUE #208: feature request: dak.to_dask_dataframe ?
I'm working on https://github.com/intake-dynamodb/intake-dynamodb/pull/4. see also https://dask.discourse.group/t/dd-from-dict-giving-typeerror/1714

What would be the best way to convert a dask_awkward.lib.core.Array to dask.DataFrame ?

Upstream there is a `ak.to_dataframe(ak_array)` method. Should there be a `dak.to_daskdataframe` method?

```
dak_array = dak.from_json("s3://bucket/AWSDynamoDB/HASH/data/*.json.gz")
ddf = dak.to_daskdataframe(dak_array) # ?
```

### PR DESCRIPTION
**Pull request by douglasdavis**
Fix #208

### CODE REVIEW

**Comment by martindurant** [comment]
OK, this is the map_partitions thing :)

**Comment by raybellwaves** [comment]
for some reason the dask.dataframe created has all types float64. I'll try and provide some code here or make a PR to your branch

Screen Shot 2023-04-04 at 12 24 26 PM

File: pyproject.toml [hunk 1]
```diff
@@ -59,6 +59,7 @@ io = [
 test = [
     "aiohttp",
     "distributed",
+    "pandas",
```
Comment (raybellwaves) [inline pyproject.toml:62]: should this be added to complete as well?

Comment (martindurant) [reply pyproject.toml:62]: I would suggest there are too many potential awkward workflows that do not need pandas for it to be even a suggested dep. If you know you will be needing pandas, awkward-pandas will eventually be the way to go.

Comment (douglasdavis) [reply pyproject.toml:62]: Yea I think we can treat it how we treat s3fs, for example. If you call `dak.from_json("s3://...")` without `s3fs` installed then there's an error message instructing you to install it.

Comment (raybellwaves) [reply pyproject.toml:62]: I think I was confused here. See https://github.com/dask-contrib/dask-awkward/issues/212

**Comment by jpivarski** [comment]
What would you think of calling this `dak.to_dataframe`, as in https://github.com/dask-contrib/dask-awkward/issues/208#issuecomment-1496480658? There's already a name distinction in `ak` vs `dak` and it should be expected that data in Dask remain in Dask until `.compute()`.

File: tests/test_io.py [hunk 2]
```diff
@@ -358,3 +358,18 @@ def test_to_json_raise_filenotfound(
...
         )
+
+
+@pytest.mark.parametrize("optimize_graph", [True, False])
+def test_to_dask_dataframe(daa: dak.Array, caa: ak.Array, optimize_graph: bool) -> None:
+    pytest.importorskip("pandas")
+
+    from dask.dataframe.utils import assert_eq
+
+    daa = daa["points", ["x", "y"]]
```
[the remark this answers is not in the document]
Comment (raybellwaves) [reply tests/test_io.py:369]: I worked on this here: https://github.com/douglasdavis/dask-awkward/pull/6

**Comment by douglasdavis** [comment]
@jpivarski looks like `ak.to_dataframe` on length zero array is losing dtype information

The reason we call `ak.to_dataframe` on the length zero array is so that we can pass a concrete length 0 dataframe as the metadata for dask.dataframe.

```python
In [23]: a = [
    ...:     {"points": [{"x": "0", "y": "9"}, {"x": "4", "y": "0"}, {"x": "4", "y": "7"}]},
    ...:     {"points": []},
    ...:     {"points": [{"x": "2", "y": "10"}, {"x": "0", "y": "10"}]},
    ...:     {"points": [{"x": "9", "y": "4"}]},
    ...:     {"points": [{"x": "2", "y": "2"}, {"x": "6", "y": "6"}, {"x": "7", "y": "9"}]},
    ...: ]

In [24]: a = ak.Array(a)

In [25]: a["points", ["x", "y"]]
Out[25]: <Array [[{x: '0', y: '9'}, ..., {...}], ...] type='5 * var * {x: string, y:...'>

In [26]: len0arr = a["points", ["x", "y"]].layout.form.length_zero_array()

In [27]: len0arr
Out[27]: <Array [] type='0 * var * {x: string, y: string}'>

In [31]: len0df = ak.to_dataframe(len0arr)

In [32]: len0df
Out[32]: 
Empty DataFrame
Columns: [x, y]
Index: []

In [33]: len0df.x
Out[33]: Series([], Name: x, dtype: float64)

In [34]: len0df.y
Out[34]: Series([], Name: y, dtype: float64)

In [35]: len0df.dtypes
Out[35]: 
x    float64
y    float64
dtype: object
```

**Comment by jpivarski** [comment]
> @jpivarski looks like `ak.to_dataframe` on length zero array is losing dtype information

This is particular to strings (and bytestrings), not all types. It happens here:

https://github.com/scikit-hep/awkward/blob/6a24ed0d436bcd158f634d9bd9f6d664fff6bd2b/src/awkward/operations/ak_to_dataframe.py#L168-L169

Aha:

```python
>>> array = ak.Array(["one", "two", "three"])

>>> array[0:0].type.show()
0 * string

>>> ak.to_numpy(array[0:0])
array([], dtype=float64)
```

The issue is in `ak.to_numpy`. scikit-hep/awkward/issues/2364

**Comment by douglasdavis** [comment]
With the fix in upstream awkward and tests passing this is good to go

**Pull request merged**